zkvm/src/predicate.rs

```diff
@@ -181,75 +151,195 @@ impl Into<CompressedRistretto> for Predicate {
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-    use bulletproofs::PedersenGens;
+impl PredicateTree {
+    /// Creates new predicate tree with a verification key and a list of programs
+    pub fn new(
+        key: Option<VerificationKey>,
+        progs: Vec<Program>,
+        blinding_key: [u8; 32],
+    ) -> Result<Self, VMError> {
+        let key = key.unwrap_or_else(|| Predicate::unsignable_key());
+        let leaves = Self::commit_leaves(&progs, blinding_key);
+        let root = MerkleTree::root(b"ZkVM.taproot", &leaves);
+
+        // P = X + h(X, M)*G
+        let adjustment_factor = Predicate::commit_taproot(key.0.as_bytes(), &root);
+        let precomputed_point = {
+            let h = adjustment_factor;
+            let x = key.0.decompress().ok_or(VMError::InvalidPoint)?;
+            let p = x + h * PedersenGens::default().B;
+            p.compress()
+        };
 
-    fn bytecode(prog: &Program) -> Vec<u8> {
-        let mut prog_vec = Vec::new();
-        prog.encode(&mut prog_vec);
-        prog_vec
+        Ok(Self {
+            leaves,
+            blinding_key,
+            key,
+            precomputed_point,
+            adjustment_factor,
+        })
     }
 
-    #[test]
-    fn valid_program_commitment() {
-        let prog = Program::build(|p| p.drop());
-        let pred = Predicate::unblinded_program(prog.clone());
-        let op = pred.prove_program_predicate(&bytecode(&prog), &[]);
-        assert!(op.verify().is_ok());
+    /// Returns the adjustment factor for signing
+    pub fn adjustment_factor(&self) -> Scalar {
+        self.adjustment_factor
     }
 
-    #[test]
-    fn invalid_program_commitment() {
-        let prog = Program::build(|p| p.drop());
-        let prog2 = Program::build(|p| p.dup(1));
-        let pred = Predicate::unblinded_program(prog);
-        let op = pred.prove_program_predicate(&bytecode(&prog2), &[]);
-        assert!(op.verify().is_err());
+    /// Creates the call proof and returns that with the program at an index.
+    pub fn create_callproof_program(
+        &self,
+        prog_index: usize,
+    ) -> Result<(CallProof, Program), VMError> {
+        let poss_leaf = &self.leaves[2 * prog_index];
+        let leaf_index = match poss_leaf {
+            PredicateLeaf::Blinding(_) => 2 * prog_index + 1,
+            PredicateLeaf::Program(_) => 2 * prog_index,
+        };
+        let tree = MerkleTree::build(b"ZkVM.taproot", &self.leaves);
+        let neighbors = tree.create_path(leaf_index).unwrap();
```

Comment:
`unwrap` is generally not good outside tests - use normal error handling over panicking.
```suggestion
        let neighbors = tree.create_path(leaf_index)?;
```